cpp/src/community/edge_triangle_count_impl.cuh

```diff
@@ -78,77 +78,298 @@ struct update_edges_p_r_q_r_num_triangles {
+          edge_first});
+      // FIXME: Consolidate both functions
+      thrust::tabulate(
+        handle.get_thrust_policy(),
+        get_dataframe_buffer_begin(vertex_pair_buffer_tmp) + intersection_indices.size(),
+        get_dataframe_buffer_begin(vertex_pair_buffer_tmp) + (2 * intersection_indices.size()),
+        extract_p_r_q_r<vertex_t, edge_t, decltype(edge_first)>{
+          prev_chunk_size,
+          1,
+          raft::device_span<size_t const>(intersection_offsets.data(), intersection_offsets.size()),
+          raft::device_span<vertex_t const>(intersection_indices.data(),
+                                            intersection_indices.size()),
+          edge_first});
+
+      thrust::sort(handle.get_thrust_policy(),
+                   get_dataframe_buffer_begin(vertex_pair_buffer_tmp),
+                   get_dataframe_buffer_end(vertex_pair_buffer_tmp));
+
+      rmm::device_uvector<edge_t> increase_count_tmp(2 * intersection_indices.size(),
+                                                     handle.get_stream());
+      thrust::fill(handle.get_thrust_policy(),
+                   increase_count_tmp.begin(),
+                   increase_count_tmp.end(),
+                   size_t{1});
+
+      auto count_p_r_q_r = thrust::unique_count(handle.get_thrust_policy(),
+                                                get_dataframe_buffer_begin(vertex_pair_buffer_tmp),
+                                                get_dataframe_buffer_end(vertex_pair_buffer_tmp));
+
+      rmm::device_uvector<edge_t> increase_count(count_p_r_q_r, handle.get_stream());
+
+      auto vertex_pair_buffer = allocate_dataframe_buffer<thrust::tuple<vertex_t, vertex_t>>(
+        count_p_r_q_r, handle.get_stream());
+      thrust::reduce_by_key(handle.get_thrust_policy(),
+                            get_dataframe_buffer_begin(vertex_pair_buffer_tmp),
+                            get_dataframe_buffer_end(vertex_pair_buffer_tmp),
+                            increase_count_tmp.begin(),
+                            get_dataframe_buffer_begin(vertex_pair_buffer),
+                            increase_count.begin(),
+                            thrust::equal_to<thrust::tuple<vertex_t, vertex_t>>{});
+
+      rmm::device_uvector<vertex_t> pair_srcs(0, handle.get_stream());
+      rmm::device_uvector<vertex_t> pair_dsts(0, handle.get_stream());
+      std::optional<rmm::device_uvector<edge_t>> pair_count{std::nullopt};
+
+      std::optional<rmm::device_uvector<edge_t>> opt_increase_count =
+        std::make_optional(rmm::device_uvector<edge_t>(increase_count.size(), handle.get_stream()));
+
+      raft::copy<edge_t>((*opt_increase_count).begin(),
+                         increase_count.begin(),
+                         increase_count.size(),
+                         handle.get_stream());
+
+      // There are still multiple copies here but is it worth sorting and reducing again?
```

Comment:
You mean after shuffle, right?

So, this is a trade-off. If you sort-and-reduce, no need to use atomics. You can experiment which will be faster.

Reply:
> You mean after shuffle, right?

Yep

> So, this is a trade-off. If you sort-and-reduce, no need to use atomics. You can experiment which will be faster

Right